CMakeLists.txt

```diff
@@ -278,6 +283,25 @@ if(USE_MKLDNN)
   set_target_properties(dnnl PROPERTIES CXX_CLANG_TIDY "")  # don't lint 3rdparty dependency
 endif()
 
+if(USE_INTGEMM)
+  message(STATUS "Using intgemm")
+  include(FetchContent)
+  FetchContent_Declare(
+    intgemm
+    GIT_REPOSITORY https://github.com/kpu/intgemm.git
+    GIT_TAG        0f05c3ebd037eacdf8cff165736fea2b0d125023
+  )
+  FetchContent_GetProperties(intgemm)
+  if(NOT intgemm_POPULATED)
+    FetchContent_Populate(intgemm)
+    add_subdirectory(${intgemm_SOURCE_DIR} ${intgemm_BINARY_DIR} EXCLUDE_FROM_ALL)
+  endif()
+  include_directories(${intgemm_SOURCE_DIR})
+  #intgemm generates a config header based on AVX512 support in the compiler.
+  include_directories(${intgemm_BINARY_DIR})
```

Comment:
It's bad practice to modify the gobal list of default include directories. If the intgemm CMakeLists.txt is correct, these two lines are not needed as `target_link_libraries(mxnet PRIVATE intgemm)` will handle the includes automatically.

Reply:
I've removed `include_directories(${intgemm_BINARY_DIR})`.  

However, `include_directories(${intgemm_SOURCE_DIR})` remains.  intgemm itself has entirely relative header paths and users do not need to `-I` the source directory to use it, so it would be incorrect to add that to the library.  

The only reason is that you asked for `FetchContent` which implies the code moves around which implies that MXNet needs some what to find the header files to begin with.  Were this to convert back to a submodule, the path to intgemm would be constant and `include_directories(${intgemm_SOURCE_DIR})` could be removed.

Reply:
> However, include_directories(${intgemm_SOURCE_DIR}) remains. intgemm itself has entirely relative header paths and users do not need to -I the source directory to use it, so it would be incorrect to add that to the library.

If users in general do not neet to `-I` the source directory, why does the particular user mxnet need to do so? Why not associate the public headers with the intgemm `target_include_directories(intgemm PUBLIC ${CMAKE_CURRENT_LIST_DIR})`?

Reply:
I've pushed a submodule version without any `include_directories`.  If you want to use `FetchContent` then the directory is variable and there needs to be some mechanism for the MXNet code to find `intgemm`.

Reply:
>  If you want to use FetchContent then the directory is variable and there needs to be some mechanism for the MXNet code to find intgemm.

Essentially you need to update the intgemm cmakelists file to ensure it encapsulates the headers as part of the intgemm target. Once you do that, cmake will handle finding intgemm header files correctly. The reason that things break once you use `FetchContent` is that the `intgemm/CMakeLists.txt` is not correct.

Currently you're hardcoding the ` #include "../../../../3rdparty/intgemm/intgemm.h"` path inside the MXNet source files. It works, but it doesn't follow the best practice.

Reply:
> Currently you're hardcoding the `#include "../../../../3rdparty/intgemm/intgemm.h"` path inside the MXNet source files.
This is common practice:
https://github.com/apache/incubator-mxnet/blob/a9ca6600384265fcf25ce2993019f9542540fda1/src/common/tensor_inspector.h#L35
https://github.com/apache/incubator-mxnet/blob/a9ca6600384265fcf25ce2993019f9542540fda1/src/operator/nn/ctc_loss.cc#L25
https://github.com/apache/incubator-mxnet/blob/a9ca6600384265fcf25ce2993019f9542540fda1/src/operator/nn/ctc_loss.cu#L27

> Essentially you need to update the intgemm cmakelists file to ensure it encapsulates the headers as part of the intgemm target. 

Were I to do that, it would unnecessarily pollute the include paths of software that references the file by its path in the source/submodule tree.

Reply:
> This is common practice:

I think the MXNet source is not always a good reference for best practices :)

> Were I to do that, it would unnecessarily pollute the include paths of software that references the file by its path in the source/submodule tree.

Having files hardcode the direct path in the submodule tree limits the use of intgemm. For example, consider users that adopt conan.io or linux distributions that prefer to package intgemm separately from MXNet.

> And it would potentially introduce header name conflicts.

It's common practice to put the public header files into `include/intgemm` instead of the top level directory. This will avoid any name conflicts.

Reply:
Header path pollution added in e6ddba887f0e96b8be288b3d3044c9a7e1531fca, now in intgemm/ directory.